tracer/src/Datadog.Trace/Activity/Handlers/DefaultActivityHandler.cs

```diff
@@ -44,21 +44,54 @@ public void ActivityStarted<T>(string sourceName, T activity)
             string? rawTraceId = null;
             string? rawSpanId = null;
 
+            // for non-IW3CActivity interfaces we'll use Activity.Id as the key as they don't have a guaranteed TraceId+SpanId
+            // for IW3CActivity interfaces we'll use the Activity.TraceId + Activity.SpanId as the key
+            // have to also validate that the TraceId and SpanId actually exist and aren't null - as they can be in some cases
+            string? activityKey = null;
+
             if (activity is IW3CActivity w3cActivity)
             {
+                if (w3cActivity.TraceId is { } activityTraceId && w3cActivity.SpanId is { } activitySpanId)
+                {
+                    activityKey = activityTraceId + activitySpanId;
+                }
+
                 // If the user has specified a parent context, get the parent Datadog SpanContext
                 if (w3cActivity.ParentSpanId is not null
-                    && w3cActivity.ParentId is string parentId
-                    && ActivityMappingById.TryGetValue(parentId, out ActivityMapping mapping))
+                 && w3cActivity.ParentId is { } parentId)
                 {
-                    parent = mapping.Scope.Span.Context;
+                    // we know that we have a parent context, but we use TraceId+ParentSpanId for the mapping
+                    // This is a result of an issue with OTel v1.0.1 (unsure if OTel or us tbh) where the
+                    // ".ParentId" matched for the Trace+Span IDs but not for the flags portion
+                    // Doing a lookup on just the TraceId+ParentSpanId seems to be more resilient
+                    if (w3cActivity.TraceId is { } && w3cActivity.ParentSpanId is { } parentSpanId)
+                    {
+                        if (ActivityMappingById.TryGetValue(w3cActivity.TraceId + w3cActivity.ParentSpanId, out ActivityMapping mapping))
+                        {
+                            parent = mapping.Scope.Span.Context;
+                        }
+                        else
+                        {
+                            // create a new parent span context for the ActivityContext
+                            var newActivityTraceId = Convert.ToUInt64(w3cActivity.TraceId.Substring(16), 16);
+                            var newActivitySpanId = Convert.ToUInt64(w3cActivity.ParentSpanId, 16);
+                            parent = Tracer.Instance.CreateSpanContext(SpanContext.None, traceId: newActivityTraceId, spanId: newActivitySpanId);
+                        }
+                    }
+                    else
+                    {
+                        // we don't have a TraceId and/or SpanId - default to the .Id
+                        if (ActivityMappingById.TryGetValue(w3cActivity.Id, out ActivityMapping mapping))
```

Comment:
Also, is there a scenario where we hit this branch? If not, let's make a note to come up with a test to exercise this separately (maybe we need to set the Id format to hierarchical)

Reply:
Ah yeah good catch, that should be searching by `ParentId`
And I'll make sure that this branch is tested

Reply:
I suggest just running the application with instrumentation and setting a breakpoint. If you don't hit it, I wouldn't spend much time right now and instead create a separate JIRA ticket to address it, as this might be a super edge case and is probably not high priority at the moment

Reply:
That branch isn't hit with the sample, so I've made a ticket to address it at a later date.